source/slang/decl-defs.h

```diff
@@ -126,6 +126,11 @@ END_SYNTAX_CLASS()
 SYNTAX_CLASS(AssocTypeDecl, AggTypeDecl)
 END_SYNTAX_CLASS()
 
+// A '__generic_param' declaration, which defines a generic 
+// entry-point parameter. Is a container of GenericTypeConstraintDecl
+SYNTAX_CLASS(GlobalGenericParamDecl, AggTypeDecl)
+END_SYNTAX_CLASS()
```

Comment:
Should this and `AssocTypeDecl` share a base class? I guess it is overkill if they aren't yet sharing any code, but it seems like they are pretty similar.

Then again, I guess a global generic parameter should *really* be trying to share code with other generic parameters, which is harder because these currently need to be aggregates.

Hmm... unless we put their `GenericTypeConstraintDecls` directly into the `ModuleDecl` instead of as children of the `GlobalGenericParamDecl` - that would match what we do for ordinary generics a bit better. To provide some context, the reason why the constraints for a generic parameter are not stored as children of the parameter are:

1) it makes it easier to enumerate all the logical parameters of the generic (and I consider constraints to be parameters, even if the user might not see them that way), because they are all stored as children of the same parent decl (even if it is a bit of a kludge right now because they don't all share a common base class...)

2) It means we can easily support generic type constraints that either come in at a different place syntactically (e.g., `void Foo<T>(...) where T : ILight { ... }`, or that don't directly apply the parameters, but rather to nested things like assocaited types (e.g., `where M.ReflectanceFunction == DisneyBRDF` would be nice to support some day)